File: flink-app/src/main/java/com/example/ctr/domain/service/EventCountAggregator.java
package com.example.ctr.domain.service;

import com.example.ctr.domain.model.Event;
import com.example.ctr.domain.model.EventCount;
import org.apache.flink.api.common.functions.AggregateFunction;

public class EventCountAggregator implements AggregateFunction<Event, EventCount, EventCount> {

    @Override
    public EventCount createAccumulator() {
        return EventCount.initial();
    }

    @Override
    public EventCount add(Event event, EventCount accumulator) {
        if (isImpressionEvent(event)) {
            accumulator.incrementImpressions();
        } else if ("click".equals(event.getEventType())) {
            accumulator.incrementClicks();
        }
        return accumulator;
    }

    private boolean isImpressionEvent(Event event) {
        String type = event.getEventType();
        return "view".equals(type) || "impression".equals(type);
    }

    @Override
    public EventCount getResult(EventCount accumulator) {
        return accumulator;
    }

    @Override
    public EventCount merge(EventCount a, EventCount b) {
        return EventCount.merge(a, b);
    }
}

File: flink-app/src/main/kotlin/com/example/ctr/domain/service/EventCountAggregator.kt
package com.example.ctr.domain.service

import com.example.ctr.domain.model.Event
import com.example.ctr.domain.model.EventCount
import org.apache.flink.api.common.functions.AggregateFunction

class EventCountAggregator : AggregateFunction<Event, EventCount, EventCount> {

    override fun createAccumulator(): EventCount = EventCount.initial()

    override fun add(event: Event, accumulator: EventCount): EventCount {
        when {
            isImpressionEvent(event) -> accumulator.incrementImpressions()
            event.eventType == "click" -> accumulator.incrementClicks()
        }
        return accumulator
    }

    private fun isImpressionEvent(event: Event): Boolean {
        val type = event.eventType
        return type == "view" || type == "impression"
    }

    override fun getResult(accumulator: EventCount): EventCount = accumulator

    override fun merge(a: EventCount, b: EventCount): EventCount = EventCount.merge(a, b)
}
